src/data/fintype/card_embedding.lean

```diff
@@ -0,0 +1,94 @@
+import data.equiv.embedding
+
+/-!
+# Birthday Problem
+
+This file establishes the cardinality of `α ↪ β` in full generality.
+-/
+
+local notation `|` x `|` := finset.card x
+local notation `‖` x `‖` := fintype.card x
+
+open_locale nat
+
+namespace fintype
+
+-- the [fintype α] argument may seem redundant; it isn't!
+lemma card_embedding_of_unique
+{α β : Type*} [unique α] [fintype α] [fintype β] [decidable_eq α] [decidable_eq β]:
+‖α ↪ β‖ = ‖β‖ := card_congr equiv.unique_embedding_equiv_result
+
+private lemma card_embedding_aux (n : ℕ) (β) [fintype β] [decidable_eq β] (h : n ≤ ‖β‖) :
+  ‖fin n ↪ β‖ = nat.desc_fac (‖β‖ - n) n :=
+begin
+  induction n with n hn,
+  { nontriviality (fin 0 ↪ β),
+    rw [nat.desc_fac_zero, fintype.card_eq_one_iff],
+    refine ⟨nonempty.some nontrivial.to_nonempty, λ x, function.embedding.ext fin.elim0⟩ },
+
+  rw [nat.succ_eq_add_one, ←card_congr (equiv.embedding_congr fin_sum_fin_equiv (equiv.refl β))],
+  rw card_congr equiv.sum_embedding_equiv_sigma_embedding_restricted,
+  all_goals { try { apply_instance } },
+
+  have : ∀ (f : fin n ↪ β), ‖fin 1 ↪ ↥((set.range f).compl)‖ = ‖β‖ - n,
+  {
+    intro f,
+    rw card_embedding_of_unique; try {apply_instance},
+    rw card_of_finset' (finset.map f finset.univ)ᶜ,
+    { simp [finset.card_compl, finset.card_map] },
+    { -- further evidence `ᶜ` is defeq, something odd
+    change ∀ x, x ∈ (finset.map f finset.univ)ᶜ ↔ x ∈ (set.range ⇑f)ᶜ,
+    simp }
+  },
+
+  rw card_sigma,
+  simp_rw this,
+  rw [finset.sum_const, finset.card_univ, nsmul_eq_mul, nat.cast_id],
+
+  set t := ‖β‖ - n.succ with ht,
+  have h' : ‖β‖ - n = t.succ,
+  { rw [ht, nat.succ_eq_add_one, ←nat.sub_sub_assoc, nat.succ_sub_one],
+    exact h,
+    exact nat.succ_pos _ },
+
+  rw [←ht, hn (nat.lt_of_succ_le h).le, h', mul_comm, nat.succ_desc_fac, nat.desc_fac_succ]
+end
+
+variables {α β : Type*} [fintype α] [fintype β] [decidable_eq α] [decidable_eq β]
+
+/- Establishes the cardinality of the type of all injections, if any exist.  -/
+@[simp] theorem card_embedding (h : ‖α‖ ≤ ‖β‖) : ‖α ↪ β‖ = (nat.desc_fac (‖β‖ - ‖α‖) ‖α‖) :=
+begin
+  trunc_cases fintype.trunc_equiv_fin α with eq,
+  rw fintype.card_congr (equiv.embedding_congr eq (equiv.refl β)),
+  exact card_embedding_aux _ _ h,
+end
+
+/-- If `‖β‖ < ‖α‖` there is no embeddings `α ↪ β`. This is the pigeonhole principle. -/
+@[simp] theorem card_embedding_eq_zero (h : ‖β‖ < ‖α‖) : ‖α ↪ β‖ = 0 :=
+begin
+  rw card_eq_zero_iff,
+  intro f,
+  obtain ⟨x, y, eq, fne⟩ := fintype.exists_ne_map_eq_of_card_lt f h,
+  have := f.injective fne, contradiction
+end
+
+theorem card_embedding_eq_if : ‖α ↪ β‖ = if ‖α‖ ≤ ‖β‖ then nat.desc_fac (‖β‖ - ‖α‖) ‖α‖ else 0 :=
+begin
+  split_ifs with h,
+    exact card_embedding h,
+    exact card_embedding_eq_zero (not_le.mp h)
+end
+
+lemma card_embedding_eq_infinite {α β} [infinite α] [fintype β] : fintype.card (α ↪ β) = 0 :=
+by erw [fintype.card, finset.univ, function.embedding.fintype', finset.card_empty]
```

Comment:
I am completely mystified by this proof! (That's not the end of the world, but how is this working?)

Reply:
I think `exact finset.card_empty` might be enough to close this goal

Reply:
Oh!

Reply:
`finset.card_empty` is `rfl` so this is actually a `rfl` lemma! I'll leave a note explaining that; this uses the `fintype.embedding'` instance, which is definitionally empty